### ISSUE TITLE
DockerCreateContainer.portBinding has no effect for plugins >= 4.1.0 (at least!)

### ISSUE THREAD

**Issue by chrisr3** [bmuschko/gradle-docker-plugin#733]
I am trying to launch a container that creates a listening port on TCP/8080, and I obviously need to be able to bind this port from the container to somewhere else on the host. To that end, I have this trivial Groovy example project:
```gradle
import com.bmuschko.gradle.docker.tasks.container.DockerCreateContainer
import com.bmuschko.gradle.docker.tasks.container.DockerStartContainer

plugins {
    id 'base'
    id 'com.bmuschko.docker-remote-api' version '4.3.0'
}

task create(type: DockerCreateContainer) {
    tty = true
    autoRemove = true
    imageId = '<My Docker image ID>'
    portBindings = ['44444:8080']

    def myBinds = [:]
    myBinds[file('config.yml').absolutePath] = '/config.xml'
    myBinds['/sys/fs/cgroup'] = '/sys/fs/cgroup'
    binds = myBinds
}

task start(type: DockerStartContainer) {
    dependsOn create
    containerId = create.containerId
}
```
I am using Gradle 5.1.1. When I invoke the wrapper using:
```
$ ./gradlew start
```
I see that my container launches correctly, _but my port-binding does not happen!_ And yet I can't see anything wrong with my `DockerCreateContainer` task either! Is there a bug in the Docker plugin, please?

**Comment by chrisr3** [comment]
At the very least, this is a documentation bug. I have managed to get this working by doing the following:
```gradle
task create(type: DockerCreateContainer) {
    tty = true
    autoRemove = true
    imageId = '<My Docker image ID>'
    portBindings = ['44444:8080']
    exposePorts('tcp', [8080])    // << NECESSARY, BUT NOT DOCUMENTED!

    def myBinds = [:]
    myBinds[file('config.yml').absolutePath] = '/app/config/config.yml'
    myBinds['/sys/fs/cgroup'] = '/sys/fs/cgroup'
    binds = myBinds
}
```

**Comment by cdancy** [comment]
So ... this is a docker thing and not a gradle (or even this plugin) thing per se. If the image itself does not expose the relevant port YOU have to manually expose it.

I will admit I've ran into this myself multiple times and it is frustrating. It might be worth doing this auto-magically for the user if we detect that portBindings were asked for but then if we do we'd have to guess the type of transport (e.g. tcp, etc..) which _may_ not be what is wanted. IDK. I'd have to think that through a bit.

**Comment by chrisr3** [comment]
Basically, the Gradle plugin's behaviour is inconsistent with the `docker` command line tool.

**Comment by cdancy** [comment]
> Basically, the Gradle plugin's behaviour is inconsistent with the docker command line tool.

This is very true but perhaps not for the reasons you understand. When on the command line you invoke something like `docker run blah blah blah` it's issuing multiple REST calls to pull the image, create the image, start the image and all sorts of other fun stuff. This plugin has tasks which are a 1-to-1 mapping on to those REST endpoints. You can think of this plugin as a very low-level building blocks type of thing that you can wire all of those commands together to create exactly what you need.

**Comment by chrisr3** [comment]
Yes, I am beginning to understand this. My problem here is that I am _unfamiliar_ with the Docker's fundamental REST API because I have only ever been exposed to Docker via its [CLI tool](https://docs.docker.com/engine/reference/commandline/cli) before.

The CLI tool ran my image successfully using:
> $ docker run -it -v \<absolute-source-path\>:/config.xml -p 44444:8080 <My-Image-ID>

And so it seemed "obvious" that I would only need to translate this command into `DockerCreateContainer` options for Gradle.

There was [nothing in the Javadoc](https://bmuschko.github.io/gradle-docker-plugin/api/com/bmuschko/gradle/docker/tasks/container/DockerCreateContainer.html) to tell me anything different :neutral_face:...

**Comment by bmuschko** [comment]
@chrisr3 Would you be interested in proposing improvements to the Javadocs as pull request?

**Comment by chrisr3** [comment]
I have just admitted that I am unfamiliar with Docker's REST API, and you have stated that your Gradle plugin _wraps_ Docker's REST API.

However, I _very much_ wish that someone who _is_ familiar with this API would document it.

**Comment by bmuschko** [comment]
@chrisr3 Thanks for your feedback. I thought it would have been a good learning experience and a way to contribute back to the project. Would you be OK with closing the issue from your end?
